src/device/ping-device.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include <ping-parser.h>
+#include <ping-port.h>
+
+/**
+ * A PingDevice class can be used to communicate with a
+ * device implementing Blue Robotics ping-protocol*
+ *
+ * *only messages from the common message set are handled by the PingDevice
+ * base class
+ *
+ * to use another message set, you should subclass PingDevice
+ * and re-implement handleMessage
+ */
+class PingDevice {
+public:
+    /**
+     *  @brief Constructor
+     *
+     *  @param port: The device io handle
+     */
+    PingDevice(PingPort& port) : _parser(4096), _port(port) {}
+
+    /**
+     *  @brief Initialize the device by establishing communications and reading in some basic data
+     *
+     *  @return true if the device was successfully initialized
+     */
+    bool initialize();
+
+    /**
+     *  @brief Read in data from device, return a ping_message pointer if available.
+     *  Data will be read in from device until there is no data left in the RX
+     *  buffer, or a valid ping_message is successfully decoded. Note that there may
+     *  still be data available in the RX buffer for decoding when this function
+     *  returns a ping_message.
+     *
+     *  @return: A pointer to the next ping_message from the device. The pointer to this message is managed
+     *  by the PingDevice object (the caller should not delete it). The pointer is only valid until the next call to
+     *  read();
+     *  @return: null if the RX buffer is empty and no ping_message has been
+     *  decoded
+     */
+    ping_message* read();
```

Comment:
this is not thread safe.

Reply:
nor is it intended to be.

patrick and I have discussed a solution to this bad pointer management at the sacrifice of performance. We will add a new function that does a copy to return the ping_message object instead of the pointer.